eslint/babel-eslint-plugin/test/rules/no-invalid-this.js

```diff
@@ -78,40 +89,94 @@ function extractPatterns(patterns, type) {
 //------------------------------------------------------------------------------
 
 const patterns = [
-  // Class private fields
   {
-    code: "class A {a = this.b;};",
-    parserOptions: { ecmaVersion: 6 },
-    valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+    code: "class A { a = this.b; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
     invalid: [],
   },
-
   {
-    code: "class A {a = () => {return this.b;};};",
-    parserOptions: { ecmaVersion: 6 },
-    valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+    code: "class A { a = () => { return this.b; }; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
     invalid: [],
   },
-
-  // Class Private methods
   {
-    code: "class A {#a = this.b;};",
-    parserOptions: { ecmaVersion: 6 },
-    valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+    code: "class A { a = () => this.b; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
     invalid: [],
   },
+  {
+    code: "class A { [this] = b; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [],
+    invalid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
+  },
+  {
+    code: "class A { [this.a] = b; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [],
+    invalid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
+  },
+  {
+    code: "class A { [`${this.a}b`] = c; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [],
+    invalid: [MODULE, SCRIPT, USE_STRICT, IMPLIED_STRICT],
+  },
+  {
+    code: "class A { [this.a] = () => { return b; }; }",
+    parserOptions: { ecmaVersion: 2020 },
+    valid: [],
```

Comment:
gloabl `this` reference is valid in `SCRIPT`, which is sloppy mode. In this case `this.a` refers to global this, so it should be valid.

Reply:
Is this true in class bodies as well? I thought class bodies were always implied strict mode (could def be wrong!).

Reply:
You are right. All parts of ClassDeclaration is strict mode. ESLint incorrectly returns `valid` for the following case
```js
class A {
  [this.x]() {}
}
```
I believe it should throw invalid this when source type is `"script"`. So in this case `SCRIPT` should be invalid, too. But we could not fix it without fixing `no-invalid-this` rule.